libASL/transforms.ml

```diff
@@ -1697,3 +1697,93 @@ module RemoveRegisters = struct
     visit_stmts v
 
 end
+
+
+
+module FixRedefinitions = struct
+  type index_name = {name: ident ; index: int}
+  type var_t =  index_name
+  module IdentMap = Map.Make(struct 
+    type t = ident
+    let compare = Stdlib.compare
+  end);;
+
+    let ident_for_v (e: var_t) : ident = 
+      if e.index = 0 then e.name else
+      match e.name with 
+      | Ident s -> Ident (s ^ "_" ^ (string_of_int e.index))
+      | FIdent (s, i) -> FIdent ((s ^ "_" ^ (string_of_int e.index), i))
+
+  class redef_renamer (globals) = object(this)
+    inherit Asl_visitor.nopAslVisitor
+
+    val mutable seen = IdentMap.empty
+    val scoped_bindings : (var_t IdentMap.t) Stack.t = 
+      let s = Stack.create () in 
+      let globals =  Seq.map (fun f -> (f, {name=f; index=0})) (IdentSet.to_seq globals) in
+      Stack.push (IdentMap.of_seq globals) s ; s
+
+    method push_scope (_:unit) : unit = Stack.push (IdentMap.empty) scoped_bindings
+    method pop_scope (_:unit) : unit = Stack.pop_opt scoped_bindings |> ignore
+    method add_bind (n: var_t) : unit = Stack.push (IdentMap.add n.name n (Stack.pop scoped_bindings)) scoped_bindings
+    method existing_binding (i: ident) : var_t option =  Stack.to_seq scoped_bindings 
+      |> Seq.find (fun s -> IdentMap.find_opt i s != None) 
+      |> Option.map (IdentMap.find i)
+
+    method incr_binding (i: ident) : var_t = 
+      let v = this#existing_binding i in
+      match v with 
+      | Some b -> {b with index = b.index + 1} 
+      | None -> {name=i; index=0}
+
+    method !vstmt s =
+      match s with 
+        | Stmt_VarDeclsNoInit(ty, vs, loc) -> 
+            let ns = List.map this#incr_binding vs in
+            List.iter this#add_bind ns;
+            ChangeTo (Stmt_VarDeclsNoInit(ty,List.map ident_for_v ns,loc))
+        | Stmt_VarDecl(ty, v, i, loc) -> 
+            let b = this#incr_binding v in 
+            this#add_bind b; ChangeTo (Stmt_VarDecl(ty, ident_for_v b, i, loc))
+        | Stmt_ConstDecl(ty, v, i, loc) ->
+            let b = this#incr_binding v in 
+            this#add_bind b; ChangeTo (Stmt_ConstDecl(ty, ident_for_v b, i, loc))
```

Comment:
if you want to override `vlvar`, you can replace the ChangeTo with a DoChildren and offload more work to the visitor. [see](https://github.com/UQ-PAC/aslp/blob/f0287df14f01477ce76d73b735ee95ad7d1b7f55/libASL/asl_visitor.ml#L309)